### PR TITLE
Single Event to Notify Players of Role Change

### PR DESCRIPTION
**Pull request by Bixkitts**
## About the PR
Coupled player notification of role changes to single point of entry (MindAddRole())

## Why / Balance
To learn more about this part of the code base, make it a little cleaner,
and to get PR #21451 merged.

## Technical details
There is now a single point of entry, MindRoleAddedEvent, raised by MindAddRole() in SharedRoleSystem.cs.
The individual RuleSystems subscribe to this and implement notification to the player however they want.
This neatly couples changing roles and notifying the player, as opposed to each RuleSystem just
calling PlayGlobal() and DispatchServerMessage() whenever they feel like it.
(nukies called it on round start, traitors called it after assigning role to mind etc.)

(Update) Just about ready to go, just need a based maintainer to point out how 90% of it is obsolete.
## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

### CODE REVIEW

**Comment by Bixkitts** [comment]
Sorry if I'm just making more work for reviewers, but I was the first to say I would try solve this.
I'm not offended at all if you ignore this and merge #21460, I'm quite a newb!

**Comment by deltanedas** [comment]
this is just a broadcast version of RoleAddedEvent with a massive amount of boilerplate needed

**Comment by Bixkitts** [comment]
All you've done so far is replace the direct calls to _sound.PlayGlobal() with adding a component that runs that function when it's added to the mind.
I'm working on pulling the notification sounds and messages out of being just wherever in the Role code and putting it in one function.
The different roles need to notify the player in different ways and fetch completely different information,
e.g. the traitor role shares the codewords and uplink code with the player upon greeting them.

That's tricky to put into one system+component and if all you're doing is calling PlayGlobal() in there on OnAdded() anyways,
 then why even have the RoleSoundSystem? Just call PlayGlobal() directly.

Your solution is good if every role just printed a basic greeting string and played a sound but that's not quite the case tho.

**Comment by Bixkitts** [comment]
And all that boilerplate is fetching the data like uplink code n stuff to notify the player with, I've just moved it into a dedicated event. 
I think nukie is good to go but traitor is still messy. WIP but the idea is you subscribe to the RoleChangeNotify event and put just
whatever logic you want in there to let the play know their role changed. 

If you can make it work with a component though that doesn't _just_ call a single function PlayGlobal() and does the other stuff related to notification too then yeah would be better I think?

**Comment by Bixkitts** [comment]
So there was a RoleBriefingComponent+System that was having a briefing string being built inside it,
but never actually printed.
Seems to have been half-implemented alongside the regular printing of messages.
Removed it.
All player notification functionality seems to cleanly 
sit in OnRoleChangeNotify() for nukie and traitors, I'll proceed with the other roles if this
looks okay.

**Comment by deltanedas** [comment]
> So there was a RoleBriefingComponent+System that was having a briefing string being built inside it,
but never actually printed.

it is printed in character menu

**Comment by Bixkitts** [comment]
That would make sense.
I grepped no reference to it except for it having a greeting string + pda stuff written to it in TraitorRuleSystem,
and a briefing string written to it in DragonRole.
Made it look like a half implemented player notify structure.
Haven't touched UI stuff at all, thanks!

**Comment by Bixkitts** [comment]
But yeah, nukie and traitor are pretty much done here
The other ones will be easy to do latest tomorrow.
TL;DR they all get a callback from MindAddRole() after calling it, specifically to notify the player of the role.

And I just realised this probably notifies all of the existing RuleSystems and not just the caller lmao.
I guess you _do_ attach a RoleNotify component of some sort to the uid getting the role?

**Comment by Bixkitts** [comment]
Yeah so now it actually plays sound in game and sends chat messages and works and stuff
I just had to read docs a few times.
NukeopsRuleComponent vs NukeopsRoleComponent vs NukeOpsComponent had
me for like the whole afternoon.
I'll cleanup and add the rest of the roles now.

**Comment by Bixkitts** [comment]
EntityQuery<>()/EntityQueryEnumerator<>() is used all over the place in the NukeopsRuleSystem to fetch the NukeopsRuleComponent but doesn't work when I call it and I don't know why.

Message log shows me it just fails to find any entities when I use it in OnRoleAddNotify(),
although NukeopsRuleComponent and GameRuleComponent entities should certainly exist by that point.

I come back every day to look at the source a little more and try find out why but I'm not seeing it yet.

**Comment by Bixkitts** [comment]
YEAH so
I was writing all my code against NukieRuleSystem and testing it by forcing nukie as admin.

Turns out that specifically forcing the nukie role as an admin doesn't create/add/start any NukieRuleComponents and so
that execution path can't use any data from the NukieRuleComponent.

My directed callback otherwise works fine and I'll have this ready tomorrow (_for sure_)

**Comment by Emisse** [comment]
I will prompt maintainer discussion on which direction to go in the case of this and #21460

**Comment by Bixkitts** [comment]
Traitor and Nukie done. (I'll do cleanup stuffs soon)
Unlike the other roles, Pirates don't even have a RoleComponent so this callback doesn't work unless I add one in!
:D

**Comment by Bixkitts** [comment]
This still doesn't pass tests because of the PirateRole serialization stuff,
but should otherwise be okay (can't test until later today).

**Comment by Bixkitts** [comment]
That seems to be all of the roles subscribed and ready to go,
gonna test and cleanup later today.
No idea what's up with the Pirate prototype stuff making the tests fail, but I'll figure it out eventually
if somebody else doesn't point it out.

**Comment by Bixkitts** [comment]
Cached the traitor uplink codes in TraitorRoleComponent, this data is unique to each traitor.

**Comment by Bixkitts** [comment]
Resolving for https://github.com/space-wizards/space-station-14/pull/21460, one sec.

**Comment by Bixkitts** [comment]
Makes little sense to use MindPlaySound() in this case, because I already resolve the player's session from the mindId and use it
to display all the messages too in one fell swoop.
MindPlaySound() would just be more indirection?

**Comment by metalgearsloth** [comment]
I agree with delta, this seems to add a lot of not really clean code and I think you're better off piecemealing this as it needs a lot of work still, sorry.
